### DISCUSSION TITLE
Use mapped path in error messages

### DISCUSSION THREAD

**Discussion by npatel-rmn** [CuyZ/Valinor#198] [Q&A]
I've been toying about with Valinor for a little a bit and am really liking it so far!

I have a use case where I've hit a hurdle; I'm using the path mapping functionality on my data source because my input data is using different field names to my PHP objects:

```php
class PackageRecord
{
  public int $ID;
  public DateTimeImmutable $startDate;
}

$input = [
  'id' => 1337,
  'start_date' => null, // invalid
];

$source = Source::array($input)
  ->map([
    'id' => 'ID',
    'start_date' => 'startDate',
  ]);

$object = (new \CuyZ\Valinor\MapperBuilder())
  ->mapper()
  ->map(PackageRecord::class, $source);
```

The issue is that when Valinor reports errors on the input fields, the node names and paths referenced in the `NodeMessage`s mention the names of my PHP object (ID, startDate). I would like to report the error messages back to the client using field names it recognises (id, start_date).

Is there an easy way to reverse the path mapping?

A super simple solution could be for me to [`array_flip`](https://www.php.net/array_flip) the mapping keys and use that to transform `ID` to `id` when looping over the error messages. However this approach breaks down when working with nested objects and path mappings. Take the example in the [docs](https://valinor.cuyz.io/latest/source/#path-mapping):

```php
$source = \CuyZ\Valinor\Mapper\Source\Source::array([
        'towns' => [
            ['label' => 'Ankh Morpork'],
            ['label' => 'Minas Tirith'],
        ],
    ])
    ->map([
        'towns' => 'cities',
        'towns.*.label' => 'name',
   ]);
```

I can't use the `towns.*.label` key verbatim, so my hope is that Valinor can help with this.

Keen to hear your thoughts and thanks for the great package!

**Comment by romm** [comment]
Hi @npatel-rmn, sorry for the late answer!

This is an interesting usecase indeed. Unfortunately, the library wont be able to help you for the time being. I didn't think about this kind of usage during the development, and to be honest I currently have no idea how to fix that.

I'll leave the discussion open, if I come with a solution at some point I'll keep you updated here. 😉

**Reply by npatel-rmn** [reply]
Thanks @romm 🎉

**Comment by TutanRamon** [comment]
I've hit the same hurdle. @npatel-rmn: have you written a workaround? If so, can you please enlighten us on your solution?

**Reply by npatel-rmn** [reply]
Unfortunately, I do not have a workaround yet. This error message mapping issue is a bit of a blocker for my needs so my usage of Valinor was halted.

**Comment by Brammm** [comment]
Sort of similarly, I've just ran into the issue where it will report the node path for an enum as `<enum>.value` if it's not a valid value, which is a little confusing as well.

**Reply by romm** [reply]
Hi @Brammm, thank you for the report. These are definitely confusing cases, for now I don't have a solution but I'll keep this open and try to think about a solution ASAP.

Have a good day. 😉

**Reply by romm** [reply]
This should now be fixed by #677 😊
